docs/cookbook/import-categories.md

```diff
@@ -0,0 +1,198 @@
+
+Fix is not that hard.
+All we need to do is to start sorting categories while importing.
+
+```diff
+ private function importCategory(array $importData): void
+ {
+     // ...
+
++    $this->categoryRepository->persistAsLastChild($category);
++    $this->em->flush();
+ }
+```
+
+> _Import finished in 20 minutes and 10 seconds_
+
+Apart the increased import time, we also had to add two new dependencies, `CategoryRepository` and `EntityManager`.
+
+## Better way
+
+One of the reasons, why the import take so long, is the category tree structure.
+After each flush, the nested set is recalculated for the tree and this take some time.
+
+But we can edit data first, and sort categories later.
+This should provide a noticeable performance improvement.
+
+At first, we remove setting parent_id in the mapping method, we will do it differently.
+
+```diff
+ private function mapImportDataToCategoryData(array $importData, CategoryData $categoryData): CategoryData
+ {
+     foreach ($this->domain->getAll() as $domainConfig) {
+         $locale = $domainConfig->getLocale();
+         $domainId = $domainConfig->getId();
+
+         $categoryData->name[$locale] = $importData['name_' . $locale];
+         $categoryData->descriptions[$domainId] = $importData['description_' . $domainId];
+     }
+
+     $categoryData->uuid = $importData['uuid'];
+-    $categoryData->parent = $importData['parent_uuid'] ? $this->categoryFacade->getByUuid($importData['parent_uuid']) : null;
+
+     return $categoryData;
+ }
+```
+
+Also, we don't need the code for sorting siblings, but we need to build the array representing the sorted adjacency list (relation to the parent).
+
+```diff
+ private array $parentIdByCategoryId;
+
+ private function importCategory(array $importData): void
+ {
+     // ...
+
+-    $this->categoryRepository->persistAsLastChild($category);
+-    $this->em->flush();
++    $parentCategoryIdOrNull = $importData['parent_uuid'] ? $this->categoryFacade->getByUuid($importData['parent_uuid'])->getId(): null;
++    $this->$parentIdByCategoryId[$category->getId()] = $parentCategoryIdOrNull;
+ }
+```
+
+As we don't change the parent and we don't sort the siblings, it's no reason to recalculate the tree and the import is much faster.
+
+> _Import finished in 7 minutes and 45 seconds_
+
+## Bring back the sorting
+
+Now we have all categories created or edited, and we have built the sorted array of ids with related parent ids.
+
+```diff
++ use Shopsys\FrameworkBundle\Model\Category\CategoryNestedSetCalculator;
+
+ public function run(): void
+ {
+     foreach ($this->getCategoryDataFromIs() as $importData) {
+         $this->importCategory($importData);
+     }
+     
++    $categoriesSortingData = CategoryNestedSetCalculator::calculateNestedSetFromAdjacencyList($this->parentIdByCategoryId);
++    $this->categoryFacade->reorderByNestedSetValues($categoriesSortingData);
+ }
+```
+
+> _Import finished in 7 minutes and 50 seconds_
```

Comment:
We have 16 mins against 20 minutes, just want to mention

Reply:
Sorry, I don't follow, can you please explain it a little bit further?